### PR TITLE
Add remaining client connection details onto event.client

### LINKED ISSUE #560: No access to tls.client
I'm trying to port some VCL to C@E, but the original has things like

```tpl
tls.client.servername # https://developer.fastly.com/reference/vcl/variables/client-connection/tls-client-servername/
tls.client.ja3_md5 # https://developer.fastly.com/reference/vcl/variables/client-connection/tls-client-ja3-md5/
```

Given the existing API, I might expect something like

```ts
import { client } from "fastly:tls";

client.servername
client.ja3_md5
```

Or perhaps chained off `FetchEvent.prototype` like `client.{address,geo}`:

```ts
fetchEvent.tls.servername
```

### PR DESCRIPTION
**Pull request by JakeChampion**
`event.client` is an object which contains information about the downstream client that made the request, currently this contains the ip address and the geolocation details

this pull-request adds the remaining client connection details which are accessible from the compute-at-edge host environment onto the `event.client` object.

the names i've chosen for the  newly added fields on `event.client` are not names i feel strongly about, and i am very happy to hear of suggestions for different names ☺️ 

resolves #560

### CODE REVIEW

File: runtime/js-compute-runtime/host_interface/host_api.h [hunk 1]
```diff
@@ -255,6 +255,18 @@ class HttpReq final : public HttpBase {
   /// Get the downstream ip address.
   static Result<HostBytes> downstream_client_ip_addr();
 
+  static Result<HostBytes> downstream_client_h2_fingerprint();
+
```
Comment (JakeChampion) [inline runtime/js-compute-runtime/host_interface/host_api.h:259]:
```suggestion
```

File: runtime/js-compute-runtime/host_interface/host_api.cpp [hunk 2]
```diff
@@ -502,6 +502,96 @@ Result<HostBytes> HttpReq::downstream_client_ip_addr() {
...
+  fastly_error_t err;
+  if (!fastly_http_req_downstream_client_h2_fingerprint(&ret, &err)) {
+    res.emplace_err(err);
+  } else {
+    res.emplace(ret);
+  }
+
+  return res;
+}
+
```
Comment (JakeChampion) [inline runtime/js-compute-runtime/host_interface/host_api.cpp:519]:
```suggestion
```

File: runtime/js-compute-runtime/host_interface/fastly.h [hunk 3]
```diff
@@ -149,28 +149,23 @@ int req_cache_override_v2_set(fastly_request_handle_t req_handle, int tag, uint3
...
-
-// (@interface func (export "downstream_tls_client_hello")
-//    (param $chello_out (@witx pointer char8))
-//    (param $chello_max_len (@witx usize))
-//    (param $nwritten_out (@witx pointer (@witx usize)))
-//    (result $err $fastly_status)
-// )
+WASM_IMPORT("fastly_http_req", "downstream_client_h2_fingerprint")
+int req_downstream_client_h2_fingerprint(char *ret, size_t ret_len, size_t *nwritten);
+
```
Comment (JakeChampion) [inline runtime/js-compute-runtime/host_interface/fastly.h:154]:
```suggestion
```

File: runtime/js-compute-runtime/fastly-world/fastly_world_adapter.cpp [hunk 4]
```diff
@@ -186,6 +186,45 @@ bool fastly_http_req_downstream_client_ip_addr(fastly_world_list_u8_t *ret, fast
       err);
 }
 
+bool fastly_http_req_downstream_client_h2_fingerprint(fastly_world_list_u8_t *ret,
+                                                      fastly_error_t *err) {
+  ret->ptr = static_cast<uint8_t *>(cabi_malloc(128, 1));
+  return convert_result(fastly::req_downstream_client_h2_fingerprint(
+                            reinterpret_cast<char *>(ret->ptr), 128, &ret->len),
+                        err);
+}
```
Comment (JakeChampion) [inline runtime/js-compute-runtime/fastly-world/fastly_world_adapter.cpp:195]:
```suggestion
```

File: runtime/js-compute-runtime/builtins/client-info.h [hunk 5]
```diff
@@ -8,6 +8,12 @@ namespace builtins {
 class ClientInfo final : public BuiltinNoConstructor<ClientInfo> {
   static bool address_get(JSContext *cx, unsigned argc, JS::Value *vp);
   static bool geo_get(JSContext *cx, unsigned argc, JS::Value *vp);
+  static bool tls_ja3_md5_get(JSContext *cx, unsigned argc, JS::Value *vp);
+  static bool tls_client_hello_get(JSContext *cx, unsigned argc, JS::Value *vp);
+  static bool tls_client_certificate_get(JSContext *cx, unsigned argc, JS::Value *vp);
+  static bool h2_fingerprint_get(JSContext *cx, unsigned argc, JS::Value *vp);
```
Comment (JakeChampion) [inline runtime/js-compute-runtime/builtins/client-info.h:14]:
```suggestion
```

File: runtime/js-compute-runtime/builtins/client-info.cpp [hunk 6]
```diff
@@ -125,6 +249,12 @@ const JSFunctionSpec ClientInfo::methods[] = {
 const JSPropertySpec ClientInfo::properties[] = {
     JS_PSG("address", address_get, JSPROP_ENUMERATE),
     JS_PSG("geo", geo_get, JSPROP_ENUMERATE),
+    JS_PSG("h2Fingerprint", h2_fingerprint_get, JSPROP_ENUMERATE),
```
Comment (JakeChampion) [inline runtime/js-compute-runtime/builtins/client-info.cpp:252]:
```suggestion
```

File: runtime/js-compute-runtime/builtins/client-info.cpp [hunk 7]
```diff
@@ -110,6 +111,129 @@ bool ClientInfo::geo_get(JSContext *cx, unsigned argc, JS::Value *vp) {
...
+    return false;
+  }
+
+  // `fingerprint` is now owned by `buffer`
+  static_cast<void>(fingerprint.ptr.release());
+
+  args.rval().setObject(*buffer);
+  return true;
+}
+
```
Comment (JakeChampion) [inline runtime/js-compute-runtime/builtins/client-info.cpp:236]:
```suggestion
```

File: integration-tests/js-compute/fixtures/client/tests.json [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+    ],
+    "downstream_request": {
+      "method": "GET",
+      "pathname": "/client/h2Fingerprint"
+    },
+    "downstream_response": {
+      "status": 200,
+      "body": "passed"
+    }
+  },
```
Comment (JakeChampion) [inline integration-tests/js-compute/fixtures/client/tests.json:40]:
```suggestion
```

File: integration-tests/js-compute/fixtures/client/bin/index.js [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+    return pass('passed')
+});
+
+routes.set("/client/h2Fingerprint", event => {
+    error = assert(event.client.h2Fingerprint instanceof ArrayBuffer, true, 'event.client.h2Fingerprint instanceof ArrayBuffer')
+    if (error) { return error }
+    error = assert(event.client.h2Fingerprint.byteLength, 39, 'event.client.h2Fingerprint.byteLength')
+    if (error) { return error }
+    return pass('passed')
+});
```
Comment (JakeChampion) [inline integration-tests/js-compute/fixtures/client/bin/index.js:27]:
```suggestion
```

**Review by guybedford — APPROVED** [review]
👍

File: runtime/js-compute-runtime/builtins/client-info.cpp [hunk 10]
```diff
@@ -125,6 +224,11 @@ const JSFunctionSpec ClientInfo::methods[] = {
 const JSPropertySpec ClientInfo::properties[] = {
     JS_PSG("address", address_get, JSPROP_ENUMERATE),
     JS_PSG("geo", geo_get, JSPROP_ENUMERATE),
+    JS_PSG("tlsCipherOpensslName", tls_cipher_openssl_name_get, JSPROP_ENUMERATE),
+    JS_PSG("tlsProtocol", tls_protocol_get, JSPROP_ENUMERATE),
+    JS_PSG("tlsJA3MD5", tls_ja3_md5_get, JSPROP_ENUMERATE),
+    JS_PSG("tlsClientCertificate", tls_client_certificate_get, JSPROP_ENUMERATE),
+    JS_PSG("tlsClientHello", tls_client_hello_get, JSPROP_ENUMERATE),
```
Comment (jamesarosen) [inline runtime/js-compute-runtime/builtins/client-info.cpp:231]: These are great!

Is it possible to add them to the `.d.ts` types file so they show up as part of `ClientInfo`?

**Pull request merged**